plugin/core/registry.py

```diff
@@ -105,8 +104,17 @@ def sessions(self, capability: Optional[str] = None) -> Generator[Session, None,
         yield from sessions_for_view(self.view, capability)
 
 
-class LspRestartClientCommand(sublime_plugin.TextCommand):
+class LspRestartServerCommand(LspTextCommand):
     def run(self, edit: Any) -> None:
         window = self.view.window()
-        if window:
-            windows.lookup(window).restart_sessions_async()
+        if not window:
+            return
+
+        config_names = [session.config.name for session in self.sessions()]
+        if config_names:
+            window.show_quick_panel(config_names, lambda index: self.restart_server(window, index, config_names))
+
+    def restart_server(self, window: sublime.Window, index: int, config_names: List[str]) -> None:
+        if index == -1:
+            return
+        windows.lookup(window).end_config_sessions_async(config_name=config_names[index])
```

Comment:
`end_config_sessions_async` ends with `_async`, so it should run in ST's worker thread. This snippet currently runs in the main thread. That means that WindowManager._sessions could be mutated while being iterated over, resulting in exceptions and angry users.

Reply:
Sorry I am not more familiar with the codebase / python. Are you meaning it should run via a `sublime.set_timeout`

Reply:
It should be run via `sublime.set_timeout_async`. That makes it run on the async thread. No need to be sorry! :) Also this is not really a python thing, it's just an ST peculiarity